File: src/main/java/io/vproxy/vfx/manager/audio/AudioGroup.java
package io.vproxy.vfx.manager.audio;

import io.vproxy.vfx.util.FXUtils;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedList;
import java.util.concurrent.ThreadLocalRandom;

public class AudioGroup {
    private static final Comparator<AudioWrapper> comparator = (a, b) -> {
        if (a.isLastPlayed()) return 1;
        if (b.isLastPlayed()) return -1;
        return a.getCount() - b.getCount();
    };
    private final LinkedList<AudioWrapper> queue = new LinkedList<>();

    public AudioGroup(AudioWrapper[] clips) {
        for (var c : clips) {
            if (c == null) continue;
            queue.add(c);
        }
        queue.sort(comparator);
    }

    public void play() {
        FXUtils.runOnFX(this::playFX);
    }

    private void playFX() {
        if (queue.size() == 0) {
            return;
        }
        if (queue.size() == 1) {
            var audio = queue.peek();
            if (audio != null) {
                audio.play();
            }
            return;
        }
        var last = queue.peekLast();
        var lastRemoved = false;
        if (last.isLastPlayed()) {
            queue.removeLast();
            last.setLastPlayed(false);
            lastRemoved = true;
        }

        var ls = new ArrayList<AudioWrapper>();
        var lastCount = -1;
        for (var a : queue) {
            if (lastCount == -1) {
                lastCount = a.getCount();
                ls.add(a);
            } else {
                if (lastCount != a.getCount()) {
                    break;
                }
                ls.add(a);
            }
        }
        var audio = ls.get(ThreadLocalRandom.current().nextInt(ls.size()));
        queue.remove(audio);
        audio.play();
        audio.setLastPlayed(true);
        if (lastRemoved) {
            queue.add(last);
        }
        queue.add(audio);
    }
}

File: src-ts/src/manager/audio/AudioGroup.ts
import { AudioWrapper } from './AudioWrapper.js';
import { FXUtils } from '../../util/FXUtils.js';

export class AudioGroup {
  private readonly queue: AudioWrapper[] = [];

  constructor(clips: Array<AudioWrapper | null>) {
    for (const c of clips) {
      if (c === null) continue;
      this.queue.push(c);
    }
    this.queue.sort(this.comparator);
  }

  private comparator = (a: AudioWrapper, b: AudioWrapper): number => {
    if (a.isLastPlayed()) return 1;
    if (b.isLastPlayed()) return -1;
    return a.getCount() - b.getCount();
  };

  play(): void {
    FXUtils.runOnFX(() => this.playFX());
  }

  private playFX(): void {
    if (this.queue.length === 0) return;
    if (this.queue.length === 1) {
      this.queue[0]!.play();
      return;
    }
    let last = this.queue[this.queue.length - 1]!;
    let lastRemoved = false;
    if (last.isLastPlayed()) {
      this.queue.pop();
      last.setLastPlayed(false);
      lastRemoved = true;
    }
    const ls: AudioWrapper[] = [];
    let lastCount = -1;
    for (const a of this.queue) {
      if (lastCount === -1) {
        lastCount = a.getCount();
        ls.push(a);
      } else {
        if (lastCount !== a.getCount()) break;
        ls.push(a);
      }
    }
    const pick = ls[Math.floor(Math.random() * ls.length)]!;
    const idx = this.queue.indexOf(pick);
    if (idx >= 0) this.queue.splice(idx, 1);
    pick.play();
    pick.setLastPlayed(true);
    if (lastRemoved) this.queue.push(last);
    this.queue.push(pick);
  }
}
